src/components/activity-client-interface.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright 2018 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AnalyticsContext} from '../proto/api_messages';
+import {ActivityIframePort} from 'web-activities/activity-ports';
+import {ActivityMode, ActivityResult} from 'web-activities/activities';
+
+/**
+ * @typedef {!Array<(string|boolean|number|null|!Array<(string|boolean|number|null)>)>} SerializedBuffer
+ */
+export class ActivityClientInterface{
+  /**
+   * @param {!HTMLIFrameElement} iframe
+   * @param {!TrustedResourceUrl} url
+   * @param {?Object=} opt_args
+   */
+  constructor(iframe, url, opt_args) {
+    this.activityIframePort_ = new ActivityIframePort(iframe, url, opt_args);
+    /** @private @const {!Object<string, function(!{SerializedBuffer})>} */
+    this.callbackMap_ = {};
+  }
+
+  /**
+   *
+   * @param {string} pub_id
+   * @param {?AnalyticsContext} context
+   */
+  start(pub_id, context) {
+    this.message(
```

Comment:
This should be implicit as part of the `ActivityPorts.openIframe` API. It's critical that we supply `publicationId` as part of connection mechanism. Otherwise we take a risk of allowing messages before security parameters are confirmed. And it'd be nice to pass the rest of the context there as well.